include/nbl/builtin/hlsl/morton.hlsl

```diff
@@ -0,0 +1,698 @@
+        }
+        NBL_IF_CONSTEXPR(Bits > 6)
+        {
+            decoded = (decoded | rightShift(decoded, 12)) & morton_mask_3_3_v<encode_t>;
+        }
+        NBL_IF_CONSTEXPR(Bits > 12)
+        {
+            decoded = (decoded | rightShift(decoded, 24)) & morton_mask_3_4_v<encode_t>;
+        }
+
+        return _static_cast<decode_t>(decoded);
+    }
+};
+
+template<uint16_t Bits, typename encode_t>
+struct MortonDecoder<4, Bits, encode_t>
+{
+    template<typename decode_t>
+    NBL_CONSTEXPR_STATIC_INLINE_FUNC decode_t decode(NBL_CONST_REF_ARG(encode_t) encodedValue)
+    {
+        arithmetic_right_shift_operator<encode_t> rightShift;
+        encode_t decoded = encodedValue & morton_mask_4_0_v<encode_t>;
+        NBL_IF_CONSTEXPR(Bits > 1)
+        {
+            decoded = (decoded | rightShift(decoded, 3)) & morton_mask_4_1_v<encode_t>;
+        }
+        NBL_IF_CONSTEXPR(Bits > 2)
+        {
+            decoded = (decoded | rightShift(decoded, 6)) & morton_mask_4_2_v<encode_t>;
+        }
+        NBL_IF_CONSTEXPR(Bits > 4)
+        {
+            decoded = (decoded | rightShift(decoded, 12)) & morton_mask_4_3_v<encode_t>;
+        }
+        NBL_IF_CONSTEXPR(Bits > 8)
+        {
+            decoded = (decoded | rightShift(decoded, 24)) & morton_mask_4_4_v<encode_t>;
+        }
+
+        return _static_cast<decode_t>(decoded);
+    }
+};
+
+// ---------------------------------------------------- COMPARISON OPERATORS ---------------------------------------------------------------
+// Here because no partial specialization of methods
+
+template<bool Signed, uint16_t Bits, uint16_t D, typename storage_t, bool BitsAlreadySpread>
+struct Equals;
+
+template<bool Signed, uint16_t Bits, uint16_t D, typename storage_t>
+struct Equals<Signed, Bits, D, storage_t, true>
+{
+    NBL_CONSTEXPR_INLINE_FUNC vector<bool, D> operator()(NBL_CONST_REF_ARG(storage_t) value, NBL_CONST_REF_ARG(vector<storage_t, D>) rhs)
+    {
+        NBL_CONSTEXPR_STATIC storage_t Mask = impl::decode_mask_v<storage_t, D, Bits>;
+        left_shift_operator<storage_t> leftShift;
+        vector<bool, D> retVal;
+        [[unroll]]
+        for (uint16_t i = 0; i < D; i++)
+        {
+            retVal[i] = (value & leftShift(Mask, i)) == leftShift(rhs[i], i);
+        }
+        return retVal;
+    }
+};
+
+template<bool Signed, uint16_t Bits, uint16_t D, typename storage_t>
+struct Equals<Signed, Bits, D, storage_t, false>
+{
+    template <typename I>
+    NBL_CONSTEXPR_INLINE_FUNC enable_if_t<is_integral_v<I>&& is_scalar_v<I> && (is_signed_v<I> == Signed), vector<bool, D> >
+    operator()(NBL_CONST_REF_ARG(storage_t) value, NBL_CONST_REF_ARG(vector<I, D>) rhs)
+    {
+        using U = make_unsigned_t<I>;
+        vector<storage_t, D> interleaved;
+        [[unroll]]
+        for (uint16_t i = 0; i < D; i++)
+        {
+            interleaved[i] = impl::MortonEncoder<D, Bits, storage_t>::encode(_static_cast<U>(rhs[i]));
+        }
```

Comment:
you're repeating the job of the constructor, and the only reason youre doing that is because you haven't defined the `operator==` out of line (not inside the class) such that you can declare the specialization of `impl::Equals` between the class (which has the ctor and forward declaration of `operator==`) and the definition of `operator==` that uses `impl::Equals